### PR TITLE
Bug: List concat unification

### LINKED ISSUE #1679: List of cells with symbolic tail not supported
Definition:
```
module LISTOFCELLS
  imports COLLECTIONS
  imports INT

  configuration <k> $PGM:KItem </k>
                <cell> <inner> 0 </inner> </cell>
                <cellList> .List </cellList>

  syntax KItem ::= "saveCell"
                 | "dropCell"

  rule <k> saveCell => . ...</k>
       <cell> CELL </cell>
       <cellList>.List => ListItem(CELL) ...</cellList>

  rule <k> dropCell => . ...</k>
       <cellList> ListItem(_) => .List  ...</cellList>

endmodule
```
Spec:
```
requires "ListOfCells.k"

module LISTOFCELLS-SPEC
  imports LISTOFCELLS

  rule <k> saveCell ~> dropCell => . </k>
       <cell> <inner> _ </inner> </cell>
       <cellList> _ </cellList>

endmodule
```
Result:
```
Failed substitution coverage check!
...
<generatedTop>
  <k>
    dropCell
  </k>
  <cell>
    <inner>
      _0
    </inner>
  </cell>
  <cellList>
    ListItem ( <cell>-fragment <inner>
      _0
    </inner> </cell>-fragment )
    _1
  </cellList>
</generatedTop>
```

Passing version of spec:
```
  rule <k> saveCell ~> dropCell => . </k>
       <cell> <inner> _ </inner> </cell>
       <cellList> .List </cellList>
```

### PR DESCRIPTION
**Pull request by andreiburdusa**
Fixes #1679 

---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

File: kore/test/Test/Kore/Builtin/List.hs [hunk 1]
```diff
@@ -258,6 +259,44 @@ test_concatAssociates =
...
+                , variableSort = listSort
+                }
+            patElemX = elementList patSymbolicX
+            patElemY = elementList patSymbolicY
+            patConcatX =
+                mkApplySymbol concatListSymbol [ patElemX, patSymbolicXs ]
+            patConcatY =
+                mkApplySymbol concatListSymbol [ patElemY, patSymbolicYs ]
+            patUnifiedXY = mkAnd patConcatX patConcatY
+        unified <- evaluateT patUnifiedXY
```
Comment (andreiburdusa) [inline kore/test/Test/Kore/Builtin/List.hs]: `unified` is always
```
\and{List{}}(
    /* term: */
    /* Fl Fn D Sfa */
    concatList{}(
        /* Fl Fn D Sfa */
        /* builtin: */ elementList{}(/* Fl Fn D Sfa */ x:Int{}),
        /* Fl Fn D Sfa */ xs:List{}
    ),
\and{List{}}(
    /* predicate: */
    /* Sfa */
    \equals{List{}, List{}}(
        /* Fl Fn D Sfa */
        concatList{}(
            /* Fl Fn D Sfa */
            /* builtin: */ elementList{}(/* Fl Fn D Sfa */ x:Int{}),
            /* Fl Fn D Sfa */ xs:List{}
        ),
        /* Fl Fn D Sfa */
        concatList{}(
            /* Fl Fn D Sfa */
            /* builtin: */ elementList{}(/* Fl Fn D Sfa */ y:Int{}),
            /* Fl Fn D Sfa */ ys:List{}
        )
    ),
    /* substitution: */
    \top{List{}}()
))
```

File: kore/test/Test/Kore/Builtin/List.hs [hunk 2]
```diff
@@ -258,6 +272,56 @@ test_concatAssociates =
...
+                        { term = patSymbolicY
+                        , predicate = makeTruePredicate listSort
+                        , substitution =
+                            Substitution.unsafeWrap
+                                [ (ElemVar elemVarX, patSymbolicY)
+                                , (ElemVar elemVarY, patSymbolicYs)
+                                ]
+                        }
+        --unified <- evaluateT patUnifiedXY
+        --traceM $ "Unified\n" <> unparseToString unified
```
Comment (andreiburdusa) [inline kore/test/Test/Kore/Builtin/List.hs]: if uncommented, this were
```
Unified
\and{List{}}(
    /* term: */
    /* Fl Fn D Sfa */ /* builtin: */ elementList{}(/* Fl Fn D Sfa */ y:Int{}),
\and{List{}}(
    /* predicate: */
    /* D Sfa */ \top{List{}}(),
    /* substitution: */
    \and{List{}}(
        /* Spa */
        \equals{Int{}, List{}}(
            /* Fl Fn D Sfa */ x:Int{},
            /* Fl Fn D Sfa */ y:Int{}
        ),
        /* Spa */
        \equals{List{}, List{}}(
            /* Fl Fn D Sfa */ xs:List{},
            /* Fl Fn D Sfa */ ys:List{}
        )
    )
))
```

File: kore/test/Test/Kore/Builtin/List.hs [hunk 3]
```diff
@@ -258,6 +291,78 @@ test_concatAssociates =
...
+                }
+            patSymbolicYs = mkElemVar elemVarYs
+            patElemX = elementList patSymbolicX
+            patElemY = elementList patSymbolicY
+            patConcatX =
+                mkApplySymbol concatListSymbol [ patElemX, patSymbolicXs ]
+            patConcatY =
+                mkApplySymbol concatListSymbol [ patElemY, patSymbolicYs ]
+            patUnifiedXY = mkAnd patConcatX patConcatY
+            expect = Conditional
```
Comment (andreiburdusa) [inline kore/test/Test/Kore/Builtin/List.hs]: `expect`
```
\and{List{}}(
    /* term: */
    /* Fl Fn D Spa */
    concatList{}(
        /* Fl Fn D Spa */ elementList{}(/* Fl Fn D Sfa */ y:Int{}),
        /* Fl Fn D Sfa */ ys:List{}
    ),
\and{List{}}(
    /* predicate: */
    /* D Sfa */ \top{List{}}(),
    /* substitution: */
    \and{List{}}(
        /* Spa */
        \equals{Int{}, List{}}(
            /* Fl Fn D Sfa */ x:Int{},
            /* Fl Fn D Sfa */ y:Int{}
        ),
        /* Spa */
        \equals{List{}, List{}}(
            /* Fl Fn D Sfa */ xs:List{},
            /* Fl Fn D Sfa */ ys:List{}
        )
    )
))
```

File: kore/test/Test/Kore/Builtin/List.hs [hunk 4]
```diff
@@ -258,6 +270,106 @@ test_concatAssociates =
...
+                        , substitution =
+                            from $ Map.fromList
+                                [ (ElemVar elemVarX, patSymbolicY)
+                                , (ElemVar elemVarXs, patSymbolicYs)
+                                ]
+                        }
+        unified <- evaluateT patUnifiedXY
+        unified' <- evaluateT patUnifiedXY'
+        expect === unified
+        expect' === unified'
```
Comment (ttuegel) [inline kore/test/Test/Kore/Builtin/List.hs:320]: I think these tests would be easier to understand if the `expect` was adjacent to the assertion itself.
```suggestion
            patConcatX = concatList patElemX patSymbolicXs
            patConcatY = concatList patElemY patSymbolicYs
            patUnifiedXY = mkAnd patConcatX patConcatY

            expect = Conditional
                        { term = patConcatY
                        , predicate = makeTruePredicate listSort
                        , substitution =
                            from $ Map.fromList
                                [ (ElemVar elemVarX, patSymbolicY)
                                , (ElemVar elemVarXs, patSymbolicYs)
                                ]
                        }
                        
        unified <- evaluateT patUnifiedXY
        expect === unified

        let
            patConcatX' = concatList patSymbolicXs patElemX
            patConcatY' = concatList patSymbolicYs patElemY
            patUnifiedXY' = mkAnd patConcatX' patConcatY'

            expect' = Conditional
                        { term = patConcatY'
                        , predicate = makeTruePredicate listSort
                        , substitution =
                            from $ Map.fromList
                                [ (ElemVar elemVarX, patSymbolicY)
                                , (ElemVar elemVarXs, patSymbolicYs)
                                ]
                        }

        unified' <- evaluateT patUnifiedXY'
        expect' === unified'
```

File: kore/src/Kore/Builtin/List.hs [hunk 5]
```diff
@@ -474,7 +490,7 @@ unifyEquals
     unifyEquals0 pat1 pat2 =
         case pat2 of
             dv@(Builtin_ (Domain.BuiltinList _)) -> unifyEquals0 dv pat1
-            _ -> empty
+            _ -> trace "\n\nHere\n\n" empty
```
[the remark this answers is not in the document]
Comment (andreiburdusa) [reply kore/src/Kore/Builtin/List.hs]: Are you sure this is part of the merged version? I can't find it

[the remark this answers is not in the document]
Comment (virgil-serbanuta) [reply kore/src/Kore/Builtin/List.hs]: Oops, sorry.

**Pull request merged**